### ISSUE TITLE
What is the correct way to use TelemetryClient with Activity?

### ISSUE THREAD

**Issue by verdysh** [microsoft/ApplicationInsights-dotnet#2604]
I have 2 apps are built on .NET 6. One of them is class library (ClassLibrary), another one asp.net core (WebApp).

ClassLibrary contains no external packages. WebApp references to ClassLibrary, also has reference to Microsoft.ApplicationInsights.AspNetCore 2.20.0. I think it's not important, but I run it on my Windows laptop.

I want my ClassLibrary publishes some telemetry for further  consumption and sending all that to Application Insights in Azure.
I added into it ActivitySource and through it I create Activities in different pieces of code. It looks like this:

```
using var activity = MyActivitySource.Current.StartActivity("MyActivity");
activity?.SetTag("tenantId", tenantId);
activity?.SetTag("onDate", onDate);
```

In my WebApp I created ActivityListener

```
_listener = new ActivityListener
{
	ShouldListenTo = source =>
	{

		return source.Name.Equals("MyActivitySourceName", StringComparison.OrdinalIgnoreCase);
	},
	Sample = (ref ActivityCreationOptions<ActivityContext> _) =>
	{
		return ActivitySamplingResult.AllData;
	},
	ActivityStarted = activity =>
	{
		
	},
	ActivityStopped = activity =>
	{
		var requestOperation = telemetryClient.StartOperation<RequestTelemetry>(activity);

		foreach (var tag in activity.Tags)
			requestOperation.Telemetry.Properties.Add(tag);

		telemetryClient.StopOperation(requestOperation);
	}
};
```

And connected it to listen my ActivitySource:

`ActivitySource.AddActivityListener(_listener);`

I get transactions in Application insights. And they are looked similar to what I wanted. But there are couple issues I don't know how to solve.

1. I don't see in Application Insights information I added into requestOperation.Telemetry.Properties.Add(tag). Neither in Request Properties nor Custom properties.
2. I have unreal time in my operation. It's too small. I don't know how can I set it correctly.

How can I fix these two issues?

**Comment by verdysh** [comment]
Hi @cijothomas 

>  2 - the duration is between the time of StartOperation and StopOperation. Which would be very low, as you are doing it in the ActivityStopped callback.

I thought I can get duration like this: `DateTime.UtcNow - activity.StartTimeUtc`.
The problem is I don't know how to set this value to `requestOperation`

> Maybe there is a better way to achieve you goal. If you can elaborate your end goal, we might be able to suggest better ways.

Let's look on SqlServerClient or HttpClient. If you add to your project reference to Microsoft.ApplicationInsights.AspNetCore and call `builder.Services.AddApplicationInsightsTelemetry()` you will get telemetry from them among other telemetry that will be viewed like a tree in Application Insights.

I want to do something similar with my library. If you connect AddApplicationInsightsTelemetry() I want you to see traces from my library in tracings inside Application Insights.

**Comment by cijothomas** [comment]
https://docs.microsoft.com/azure/azure-monitor/app/asp-net-dependencies#manually-tracking-dependencies - To manually track dependencies not automatically collected.

**Comment by verdysh** [comment]
If I use this approach in my library, my library will be addicted to Microsoft.ApplicationInsights.AspNetCore package. My idea is exactly to avoid that. I want to give an opportunity to library users to get some diagnostic information (traces) from my library, connected with distributed tracing as well. But the choice how will you use this information I want leave to user.

Maybe you want just write some flatten logs into you system if you have monolith app and distributed tracing isn't needed in your case. Maybe you want exporting this traces to some tracing system and Application Insights it's only one option among others.

If you want doing nothing with this information you can doing nothing but **library code stays the same regardless do you use this traces or not**. What is bolded this is the main reason why I thought to use approach with Activity / ActivityListener. I thought this is exactly what I need.

**Comment by verdysh** [comment]
> setup own listener

Do you mean ActivityListener like in my example above or something else?

> call ApplicationInsights specific API (like TrackDependency etc.).

I'm calling StartOperation() and StopOperation() because I thought as StartOperation method accepts an Activity object, this is exactly what I need. The only problem leaved it's I couldn't set correct time. I have this time but I don't know where I have to pass it. I tried the next way:

```
var requestOperation = telemetryClient.StartOperation<RequestTelemetry>(activity);
requestOperation.Telemetry.Duration = activity.Duration; // it doesn't work
telemetryClient.StopOperation(requestOperation);
```

But it didn't work. If I resolve this issue I will be happy.

PS. issue with tags I resolved by moving necessary tags from SetTag to SetBaggage method; with SetBaggage it works.

**Comment by cijothomas** [comment]
You are better off doing TrackDependency/TrackRequest in the OnActivityEnd callback.


// following shows hardcode values. but they can come from Activity and Activity tags.
        DependencyTelemetry dep = new DependencyTelemetry();
        dep.Name = "DependencyName";
        dep.Data = "SqlStatement";
        dep.Type = "Sql";
        dep.Target = "TargetServer";
        dep.Duration = TimeSpan.FromSeconds(10);
        dep.ResultCode = "500";
        dep.Success = false;

        dep.Context.Cloud.RoleName = "MyRole";
        dep.Context.Cloud.RoleInstance = "MyRoleInstance";
        client.TrackDependency(dep);

I'd also suggest to not use SetBaggage. It serves different purpose.

**Comment by verdysh** [comment]
> You are better off doing TrackDependency/TrackRequest in the OnActivityEnd callback.

Okay, thanks!

> I'd also suggest to not use SetBaggage. It serves different purpose.

I have an issue with tags. I tried
```
activity.AddTag("bbb", 2);
activity.SetTag("bbbbbb", 22);
```
but as a result I see anything in activity.Tags
Could you tell please a little bit detailed when I should SetBaggage(), when and how I should use xxxTag() ?

**Comment by cijothomas** [comment]
> but as a result I see anything in activity.Tags

Could you elaborate this? What exactly is the issue with Activity.SetTag

**Comment by verdysh** [comment]
Here is part of how is looked my end-to-end transactions in Azure: https://ibb.co/z4QCV33

I marked with a red rectangle my custom events which were created with activity and which were exported by my activity listener.
Area marked with green rectangle it's what I want to see only inside root operation. In my case I want to see that only inside PartnerCustomersReportUploadingService.Uploading operation.

When I used tags nothing happened. I didn't see these properties in end-to-end transactions. I even didn't see added tags inside activity.Tags property when I'm debugging.

When I'm using SetBaggage I see these properties inside end-to-end transactions. What I really wanted is to see them only inside root operation for which I set them. Currently I see them inside all downstream items including logs items. For sure it's better than not seeing them at all. That's why I settled on using Baggage.
